pkg/handler/unrouted_handler.go

```diff
@@ -650,14 +662,24 @@ func (handler *UnroutedHandler) HeadFile(w http.ResponseWriter, r *http.Request)
 
 		resp.StatusCode = http.StatusOK
 	} else {
-		if !info.SizeIsDeferred && info.Offset == info.Size {
-			// Upload is complete if we know the size and it matches the offset.
-			resp.Header["Upload-Complete"] = "?1"
-		} else {
-			resp.Header["Upload-Complete"] = "?0"
+		currentUploadDraftInteropVersion := handler.getResumeableUploadDraftVersion(r)
+		uploadComplete := !info.SizeIsDeferred && info.Offset == info.Size
+		if currentUploadDraftInteropVersion == Version4 {
+			if uploadComplete {
+				resp.Header["Upload-Complete"] = "?1"
+			} else {
+				resp.Header["Upload-Complete"] = "?0"
+			}
+		} else if currentUploadDraftInteropVersion == Version3 {
+			if uploadComplete {
+				// Upload is complete if we know the size and it matches the offset.
+				resp.Header["Upload-Incomplete"] = "?0"
+			} else {
+				resp.Header["Upload-Incomplete"] = "?1"
+			}
```

Comment:
The logic of setting Upload-Complete or Upload-Incomplete based on the interop version can be refactored into its own function. That should make the code a bit easier to read and avoid duplication.

Reply:
I'm finding it difficult to make a version of this that is cleaner without introducing indirection that is hard to follow. I made some minor improvements (and deduplicated code elsewhere) — would love your thoughts.

Reply:
No worries, I just did it on my own. Have a look at https://github.com/tus/tusd/pull/1066/commits/1e4cd41656da7c0925ec8f2ebd6d7d94fe5a432b to see what I meant.